util/topgen/secure_prng.py

```diff
@@ -0,0 +1,289 @@
+                new_seed <<= seed_bytes * 8
+                new_seed += seed
+            new_seed %= (1 << 256)
+            log.warning("Seed smaller than 256 bits. CTR_DRBG seeded with" +
+                        hex(new_seed))
+            entropy_input = new_seed.to_bytes(32, 'big')
+        else:
+            entropy_input = seed.to_bytes(32, 'big')
+        self.CTR_DRBG_Instantiate(entropy_input)
+        self.returned_bits = []
+
+    def fetchbyte(self):
+        """Fetches the next byte from the returned_bits.
+
+        Used by getrandbits().
+        If the returned_bits are consumed, requests fresh 512 bits from CTR_DRBG
+        """
+
+        if not self.returned_bits:
+            self.returned_bits = list(self.CTR_DRBG_Generate(512))
+
+        return self.returned_bits.pop(0)
+
+    def getrandbits(self, n_bits):
+        """Fetches n_bits next bits from a PRNG.
+
+        """
+
+        bitsleft = n_bits
+        R = 0
+        while (bitsleft > 0):
+            random_byte = int(self.fetchbyte())
+            if (bitsleft > 8):
+                R = (R << 8) | random_byte
+                bitsleft = bitsleft - 8
+            else:
+                random_byte >>= 8 - bitsleft
+                R = (R << bitsleft) | random_byte
+                bitsleft = 0
+        assert R >= 0
+        return (R)
+
+    def randbelow(self, n):
+        """Generates a random integer smaller than n.
+
+        Uses getrandbits() to generate enough random bits. Repeats until the
+        resulting number is smaller than n. The probability of this is more
+        than 50% in each loop iteration.
+        The amount of consumed randomness is not fixed.
+        """
+        length_bits = _ceil(_log(n, 2))
+        R = self.getrandbits(length_bits)
+        while R >= n:
+            R = self.getrandbits(length_bits)
+        return (R)
+
+    def shuffle(self, x):
+        """Shuffles a list x.
+
+        Uses the Fisher-Yates algorithm.
+        Each possible permutation of x is generated with equal probability.
+        """
+
+        length_list = len(x)
+        for i in reversed(range(length_list)):
+            j = self.randbelow(i + 1)
+            x[i], x[j] = x[j], x[i]
+
+    def choice(self, x):
+        """Randomly chooses an element from a list x.
+
+        Uses randbelow() to select an integer smaller than len(x) and returns
+        the corresponding element.
+        """
+
+        i = self.randbelow(len(x))
+        return (x[i])
+
+    def test_print_state(self):
+        print("Key = ", f"{self.Key:#0{34}x}")
+        print("V   = ", f"{self.V:#0{34}x}")
+
+    def unit_test(self, entropy_input, verbose = False):
+        self.reseed(entropy_input)
+        if verbose:
+            print("Entropy Input = ", hex(entropy_input))
+            print("INSTANTIATE: ")
+            self.test_print_state()
+        K1 = self.Key
+        V1 = self.V
+        test_word_bytes = self.CTR_DRBG_Generate(512)
```

Comment:
This is intentional.
The test vectors make two calls to `CTR_DRBG_Generate`, and only output the numbers of the second call. So we don't need the `ReturnedBits` after the first call because we have nothing to compare them with.

Reply:
Okay thanks for clarifying.